### PR TITLE
Join trip from invite link 

### PR DESCRIPTION
**Pull request by yongbingao**
@kevinlin505 @wleofeng take a look

### CODE REVIEW

File: src/components/TripDetail/TripDetail.js [hunk 1]
```diff
@@ -1,7 +1,39 @@
...
+};
+
+const TripDetail = ({ auth, actions, match }) => {
+  useEffect(() => {
+    const currentTripId = match.params.tripId;
+    let inTrip = false;
+    auth.profile.trips.forEach(el => {
+      if (el.id === currentTripId) inTrip = true;
+    });
+    if (!inTrip) {
```
Comment (kevinlin505) [inline src/components/TripDetail/TripDetail.js]: Nit: maybe have a line break above if for better readability

File: src/components/TripDetail/TripDetail.js [hunk 2]
```diff
@@ -1,7 +1,39 @@
...
+      actions.user.updateProfile({ joinTripId: currentTripId });
+      actions.trip.updateTrip({ tripId: currentTripId });
+    }
+  }, []);
```
Comment (kevinlin505) [inline src/components/TripDetail/TripDetail.js]: This part need to restructure. For the intrip variable it should be a state instead and separate its logic to a separate useEffect

File: src/components/TripDetail/TripDetail.js [hunk 3]
```diff
@@ -1,7 +1,39 @@
...
+
+const TripDetail = ({ auth, actions, match }) => {
+  useEffect(() => {
+    const currentTripId = match.params.tripId;
+    let inTrip = false;
+    auth.profile.trips.forEach(el => {
+      if (el.id === currentTripId) inTrip = true;
+    });
+    if (!inTrip) {
+      actions.user.updateProfile({ joinTripId: currentTripId });
```
Comment (kevinlin505) [inline src/components/TripDetail/TripDetail.js]: I am seeing the use of inviteTripId and joinTripId here, should be make them the same or is it different in this case.

File: src/components/UserPage/UserPage.js [hunk 4]
```diff
@@ -32,6 +34,20 @@ const UserPage = ({ actions, auth, trip }) => {
     setCreateTripModalOpen(!isCreateTripModalOpen);
   };
 
+  useEffect(() => {
+    if (trip.inviteTripId) {
+      let inTrip = false;
+
+      auth.profile.trips.forEach(el => {
+        if (el.id === trip.inviteTripId) inTrip = true;
+      });
```
Comment (kevinlin505) [inline src/components/UserPage/UserPage.js]: We should update this in mapStateToProp using a helper

File: src/data/user.js [hunk 5]
```diff
@@ -45,10 +45,16 @@ export default function user() {
     },
 
     updateUserProfile: (profile = {}) => {
+      const newProfile = profile;
+      if (newProfile.joinTripId) {
+        const tripRef = db.collection('trips').doc(newProfile.joinTripId);
+        newProfile.trips = firebase.firestore.FieldValue.arrayUnion(tripRef);
+        delete newProfile.joinTripId;
```
Comment (kevinlin505) [inline src/data/user.js]: Why not just use profile?

File: src/components/TripDetail/TripDetail.js (hunk 2 above)
Comment (kevinlin505) [reply src/components/TripDetail/TripDetail.js]: Per my below comment, I think the intrip logic with be a helper for MapStateToProp

File: src/components/TripDetail/TripDetail.js (hunk 3 above)
Comment (yongbingao) [reply src/components/TripDetail/TripDetail.js]: Changed to joinTripId and references

**Comment by yongbingao** [comment]
@kevinlin505 take a look at the files, it seem to work but I'm not exactly sure how the createSelectorCreator works

File: src/components/TripDetail/TripDetail.js [hunk 6]
```diff
@@ -1,36 +1,43 @@
...
 import { userActions } from '@providers/user/user';
 import { tripActions } from '@providers/trip/trip';
+import getTripStatus from '@selectors/trip_selector';
 
-const mapStateToProps = state => {
+const mapStateToProps = (state, ownProps) => {
+  const joinTripId = ownProps.match.params;
   return {
     auth: state.auth,
+    userInTrip: getTripStatus(state, joinTripId),
```
Comment (kevinlin505) [inline src/components/TripDetail/TripDetail.js]: no need to do ownProps, see below
```
const mapStateToProps = (state, props) => {
  return {
    auth: state.auth,
    tripId: props.match.params.tripId,
    userInTrip: getTripStatus(state, props),
  };
};
```

File: src/components/TripDetail/TripDetail.js [hunk 7]
```diff
@@ -1,36 +1,43 @@
...
+  // }, []);
+
   useEffect(() => {
-    const currentTripId = match.params.tripId;
-    let inTrip = false;
-    auth.profile.trips.forEach(el => {
-      if (el.id === currentTripId) inTrip = true;
-    });
-    if (!inTrip) {
+    if (userInTrip === false) {
```
Comment (kevinlin505) [inline src/components/TripDetail/TripDetail.js]: just do `if (! userInTrip )`

File: src/components/TripDetail/TripDetail.js [hunk 8]
```diff
@@ -1,36 +1,43 @@
...
-      user: bindActionCreators(userActions, dispatch),
       trip: bindActionCreators(tripActions, dispatch),
+      user: bindActionCreators(userActions, dispatch),
     },
   };
 };
 
-const TripDetail = ({ auth, actions, match }) => {
+const TripDetail = ({ actions, match, userInTrip }) => {
+  const currentTripId = match.params.tripId;
```
Comment (kevinlin505) [inline src/components/TripDetail/TripDetail.js]: just use the tripId from the msp.

File: src/components/UserPage/UserPage.js [hunk 9]
```diff
@@ -35,15 +35,17 @@ const UserPage = ({ actions, auth, trip }) => {
...
   useEffect(() => {
-    if (trip.inviteTripId) {
+    const { joinTripId } = trip;
+
+    if (joinTripId) {
       let inTrip = false;
 
       auth.profile.trips.forEach(el => {
-        if (el.id === trip.inviteTripId) inTrip = true;
+        if (el.id === joinTripId) inTrip = true;
```
Comment (kevinlin505) [inline src/components/UserPage/UserPage.js]: Do the samething here with `userInTrip`

File: src/selectors/trip_selector.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import createDeepEqualSelector from '@utils/cache_helper';
+
+const getTripList = state => state.auth.profile.trips;
+
+const getJoinTripId = joinTripId => joinTripId;
```
Comment (kevinlin505) [inline src/selectors/trip_selector.js]: `const getJoinTripId = state => state.trip.joinTripId;`

File: src/selectors/trip_selector.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+const getTripStatus = createDeepEqualSelector(
+  getTripList,
+  getJoinTripId,
+  (tripList, joinTripId) => {
+    return tripList.some(el => {
+      return el.id === joinTripId;
+    });
+  },
+);
```
Comment (kevinlin505) [inline src/selectors/trip_selector.js]:
```
const getTripStatus = createDeepEqualSelector(
  [getTripList, getJoinTripId],
  (tripList, joinTripId) => tripList.some(el => {
      return el.id === joinTripId;
    }),
);
```

File: src/selectors/trip_selector.js [hunk 12]
```diff
@@ -0,0 +1,17 @@
+import createDeepEqualSelector from '@utils/cache_helper';
```
Comment (kevinlin505) [inline src/selectors/trip_selector.js]: let's call it `cacheHelper` instead. Also, try to use camelCase for the filename.

**Comment by yongbingao** [comment]
@kevinlin505 check it out

File: src/components/TripDetail/TripDetail.js [hunk 13]
```diff
@@ -24,20 +23,11 @@ const mapDispatchToProps = dispatch => {
...
-  // }, []);
-
+const TripDetail = ({ actions, tripId, userInTrip }) => {
   useEffect(() => {
-    if (userInTrip === false) {
-      actions.user.updateProfile({ joinTripId: currentTripId });
-      actions.trip.updateTrip({ joinTripId: currentTripId });
+    if (!userInTrip) {
+      actions.user.updateProfile({ joinTripId: tripId });
+      actions.trip.updateTrip({ joinTripId: tripId });
```
Comment (kevinlin505) [inline src/components/TripDetail/TripDetail.js]: 👍

File: src/selectors/tripSelector.js [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import createDeepEqualSelector from '@utils/cacheHelper';
+
+const getTripList = state => state.auth.profile.trips;
+
+const getJoinTripId = state => state.trip.joinTripId;
+
+const getTripStatus = createDeepEqualSelector(
+  [getTripList, getJoinTripId],
+  (tripList, joinTripId) => tripList.some(el => el.id === joinTripId),
+);
```
Comment (kevinlin505) [inline src/selectors/tripSelector.js]: 💣 🔥

**Pull request merged**